src/client-side-metrics/metrics-tracer-factory.ts

```diff
@@ -0,0 +1,600 @@
+   */
+  private getOperationLatencyAttributes(
+    projectId: string,
+    finalOperationStatus: string,
+    streamOperation?: string
+  ): Attributes {
+    return Object.assign(
+      {
+        finalOperationStatus: finalOperationStatus,
+        streamingOperation: streamOperation,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for final operation metrics. These attributes provide
+   * context about the Bigtable environment and the operation being performed.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} finalOperationStatus The final status of the operation.
+   * @returns {Attributes} An object containing the attributes for final operation metrics.
+   */
+  private getFinalOpAttributes(
+    projectId: string,
+    finalOperationStatus: string
+  ): Attributes {
+    return Object.assign(
+      {
+        finalOperationStatus: finalOperationStatus,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for attempt metrics. These attributes provide context
+   * about the Bigtable environment, the operation being performed, and the status of the attempt.
+   * Includes whether the operation was a streaming operation or not.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} attemptStatus The status of the attempt.
+   * @param {string} streamingOperation Whether the operation was a streaming operation or not.
+   * @returns {Attributes} An object containing the attributes for attempt metrics.
+   */
+  private getAttemptAttributes(
+    projectId: string,
+    attemptStatus: string,
+    streamingOperation: string
+  ) {
+    return Object.assign(
+      {
+        attemptStatus: attemptStatus,
+        streamingOperation: streamingOperation,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for attempt status metrics. These attributes provide context
+   * about the Bigtable environment and the operation being performed.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} attemptStatus The status of the attempt.
+   * @returns {Attributes} An object containing the attributes for attempt status metrics.
+   */
+  private getAttemptStatusAttributes(projectId: string, attemptStatus: string) {
+    return Object.assign(
+      {
+        attemptStatus: attemptStatus,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Called when the operation starts. Records the start time.
+   */
+  onOperationStart() {
+    this.operationStartTime = this.dateProvider.getDate();
+  }
+
+  /**
+   * Called after the client reads a row. Records application blocking latencies.
+   */
+  onRead() {
+    const currentTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (this.lastReadTime) {
```

Comment:
I assume this is doing different logic for the first row vs subsequent rows

It might be worth doing something a bit more complex here, to make sure assumptions hold (i.e. a state machine or something similar)

The state can get complicated, and we want to be careful that we don't accidentally send wrong data